design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+
+1. User sets deletion timestamp on all resources (that automatically triggers
+  the deletion of the Provider Switch because of OwnerRefs).
+2. Switch deletion is blocked until all SwitchPorts are deleted.
+3. Wait for switch port to be idle.
+4. If switchPort is idle, then remove the finalizer.
+5. Once no ports are left for the Switch, then remove the finalizer on
+  the ProviderSwitch and on the Switch.
+
+### Changes to Current API
+
+#### Metal3Machine CRD
+
+```yaml
+spec:
+  # Specify the network configuration that metal3Machine needs to use
+  networkConfiguration:
+  # no smartNic
+  - ConfigurationRef:
+    name: nc1
+    kind: SwitchPortConfiguration
+    namespace: default
+    # Network card performance required for this network configuration
+    nicHint:
+      name: eth0
+      smartNIC: false
+  # with smartNic
+  - ConfigurationRef:
+    name: nc1
+    kind: SmartNicConfiguration
+    namespace: default
+    nicHint:
+      name: eth1
+      smartNIC: true
+```
+
+#### BareMetalHost
+
+Add the `ports` field to `.spec` to indicate the connection information and
+performance of all network cards in the `BareMetalHost`. The name field is
+provided for direct matching with `Metal3Machine` network configuration,
+in case the hardware details are not provided, and the mac address is
+provided to match the introspection data, when nicHints are provided
+and the name is not.
+
+```yaml
+  spec:
+    ports:
+    - mac: 00:00:00:00:00:00
+      port:
+        name: port0
+        kind: SwitchPort
+        namespace: default
+      provisioningConfiguration:
+        name: provisioningSwitchPortConfiguration
+        kind: SwitchportConfiguration
+        namespace: default
+      configuration:
+        name: switchPortConfiguration
+        kind: SwitchportConfiguration
+        namespace: default
+```
+
+### Add new CRD and controller
+
+#### Switch
+
+The following is the realization of the switch:
+
+##### ProviderSwitch CRD
+
+ProviderSwitch CRDs include the login information of switches for different
+protocols. For each ProviderSwitch CRD, there will be a provider
+implementation knowing how to interact with the switch with this information.
+
+###### AnsibleSwitch CRD
+
+Use ansible as the backend to connect to the configuration switch
+(using the network-runner tool, currently only supports the switches supported
+by network-runner).
+
+```yaml
+apiVersion: v1alpha1
+kind: AnsibleSwitch
+metaData:
+  name: sp0
+  namespace: default
+spec:
+  # The type of OS this switch runs
+  os: fos
+  # The ip address of the switch
+  ip: 192.168.0.1
+  # The port to use for SSH connection
+  port: 22
+  # Only for ovs switch
+  bridge:
+  # The login credentials of switch
+  secret:
+    name: switch-example-secret
+    namespace: default
+```
+
+##### Switch CRD
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: Switch
+metadata:
+  name: switch-example
+spec:
+  provider:
+    kind: AnsibleSwitch
+    name: ansibleswitch-sample
```

Comment:
Would it be reasonable to assume that the provider CR has to have the same name as the Switch CR?

Reply:
Is it necessary to do this? I think it's strange that different CRDs have the same name.